wdl-engine/src/http/azure.rs

```diff
@@ -49,46 +51,56 @@ pub(crate) fn rewrite_url(url: &Url) -> Result<Url> {
 
 /// Applies Azure SAS token authentication to the given URL.
 ///
-/// Returns `true` if the URL was for Azure or `false` if it was not.
-pub(crate) fn apply_auth(config: &AzureStorageConfig, url: &mut Url) -> bool {
-    if let Some(url::Host::Domain(domain)) = url.host() {
-        if let Some(account) = domain.strip_suffix(AZURE_STORAGE_DOMAIN_SUFFIX) {
-            // If the URL already has a query string, don't modify it
-            if url.query().is_some() {
-                return true;
-            }
-
-            if let Some(mut segments) = url.path_segments() {
-                // Determine the container name; if there's only one path segment, then use the
-                // root container name
-                let container = match (segments.next(), segments.next()) {
-                    (Some(_), None) => ROOT_CONTAINER_NAME,
-                    (Some(container), Some(_)) => container,
-                    _ => return true,
-                };
-
-                if let Some(containers) = config.auth.get(account) {
-                    if let Some(token) = containers.get(container) {
-                        // Warn if the scheme isn't https, as we won't be applying the auth.
-                        if url.scheme() != "https" {
-                            warn!(
-                                "Azure Blob Storage URL `{url}` is not using HTTPS: \
-                                 authentication will not be used"
-                            );
-                            return true;
-                        }
-
-                        let token = token.strip_prefix('?').unwrap_or(token);
-                        url.set_query(Some(token));
-                    }
-                }
-            }
-
-            return true;
+/// Returns `(false, _)` if the URL is not for Azure Blob Storage.
+pub(crate) fn apply_auth<'a>(
+    config: &AzureStorageConfig,
+    url: Cow<'a, Url>,
+) -> (bool, Cow<'a, Url>) {
```

Comment:
What do these represent? Maybe a quick doc comment on the return types meaning here.

Reply:
(I see their usage in context, but I still think documenting here is prudent)

Reply:
Yeah, it's a bit wonky as the downloader is trying its best not to allocate for a HTTP/S URL that requires no auth.

I'll expand on the comments.